eth/beacon/deposit_helpers.py

```diff
@@ -0,0 +1,160 @@
+from typing import (
+    Sequence,
+    Tuple,
+)
+
+from eth_typing import (
+    Hash32,
+)
+from eth_utils import (
+    ValidationError,
+)
+
+from eth.utils import bls
+
+from eth.beacon.constants import (
+    EMPTY_SIGNATURE,
+)
+from eth.beacon.enums import (
+    SignatureDomain,
+)
+from eth.beacon.exceptions import (
+    MinEmptyValidatorIndexNotFound,
+)
+from eth.beacon.types.deposit_input import DepositInput
+from eth.beacon.types.states import BeaconState
+from eth.beacon.types.validator_records import ValidatorRecord
+from eth.beacon.helpers import (
+    get_domain,
+)
+
+
+def get_min_empty_validator_index(validators: Sequence[ValidatorRecord],
+                                  current_slot: int,
+                                  zero_balance_validator_ttl: int) -> int:
+    for index, validator in enumerate(validators):
+        if (
+                validator.balance == 0 and
+                validator.latest_status_change_slot + zero_balance_validator_ttl <=
+                current_slot
+        ):
+            return index
+    raise MinEmptyValidatorIndexNotFound()
+
+
+def validate_proof_of_possession(state: BeaconState,
+                                 pubkey: int,
+                                 proof_of_possession: bytes,
+                                 withdrawal_credentials: Hash32,
+                                 randao_commitment: Hash32) -> None:
+    deposit_input = DepositInput(
+        pubkey=pubkey,
+        withdrawal_credentials=withdrawal_credentials,
+        randao_commitment=randao_commitment,
+        proof_of_possession=EMPTY_SIGNATURE,
+    )
+
+    is_valid_signature = bls.verify(
+        pubkey=pubkey,
+        # TODO: change to hash_tree_root(deposit_input) when we have SSZ tree hashing
+        message=deposit_input.root,
+        signature=proof_of_possession,
+        domain=get_domain(
+            state.fork_data,
+            state.slot,
+            SignatureDomain.DOMAIN_DEPOSIT,
+        ),
+    )
+
+    if not is_valid_signature:
+        raise ValidationError(
+            "BLS signature verification error"
+        )
+
+
+def add_pending_validator(state: BeaconState,
+                          validator: ValidatorRecord,
+                          zero_balance_validator_ttl: int) -> Tuple[BeaconState, int]:
+    """
+    Add a validator to the existing minimum empty validator index or
+    append to ``validator_registry``.
+    """
+    # Check if there's empty validator index in `validator_registry`
+    try:
+        index = get_min_empty_validator_index(
+            state.validator_registry,
+            state.slot,
+            zero_balance_validator_ttl,
+        )
+    except MinEmptyValidatorIndexNotFound:
+        index = None
+
+    if index is None:
```

Comment:
I think this would all be a little better readability if you had all of this in the `try/except/else`

```python
try:
    index = ...
except MinEmptyValidatorIndexNotFound:
    ...  # logic from the `if index is None`
else:
    ...  # other logic.

return state, index
```